---
title: "DADA2 pipeline"
author: "Stephen Noell"
date: "21/06/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      message = FALSE, 
                      warning = FALSE,
                      fig.align = "center",
                      fig.width = 10,
                      fig.height = 6)
```


```{r}

# Libraries
library("dada2")
library(ShortRead)
library(Biostrings)
library(seqinr)
library(DECIPHER)
library(readr)
library(phangorn)
library("tidyverse")       # data wrangling and visualisation
library("phyloseq")        # analysis of microbial communities
library("here")            # set the path to the folder
library("ggplot2")
library("ggpubr") #for putting figs together
library("RColorBrewer") #for color palettes
library("svglite") #for saving SVGs

set.seed(57)
```


```{R}
#ASV table read in
seqtab_nochim <- read.csv("seqtab_nochim_es2.csv") %>%
  column_to_rownames(var = "X") %>%
  as.matrix(.)
  
# Assign taxonomy ---- using DADA2 bayesian classifier

# DADA2 provides a native implementation of the naive Bayesian classifier method for this
# Maintains formatted training fastas - https://benjjneb.github.io/dada2/training.html

# Using the naive bayesian classifier ----

# Silva database; using v. 138.2
taxa.silva <- assignTaxonomy(seqtab_nochim, "C:/Users/snoell/OneDrive - The University of Waikato/Documents/silva_nr99_v138.2_toGenus_trainset.fa.gz", multithread=FALSE)

# Save the taxanomy table
saveRDS(taxa.silva, "es_tax_table_silva")

#es_tax <- as.data.frame(readRDS("es_tax_table"))
#taxa.silva <- as.data.frame(readRDS("es_tax_table_silva"))

```


```{r}
# Tree ----
# Get the seqs and store the them so they can be the tips of the tree
es_seqs <- getSequences(seqtab_nochim)
names(es_seqs) <- es_seqs
# Align sequences (using DECIPHER)
alignment <- AlignSeqs(DNAStringSet(es_seqs), anchor = NA,verbose = FALSE)

# Build a tree (using phangorn)
phangAlign <- phyDat(as(alignment, "matrix"), type = "DNA")
dm <- dist.ml(phangAlign)
treeNJ <- NJ(dm) 
fit = pml(treeNJ, data=phangAlign)
fitGTR <- update(fit, k=4, inv=0.2)
fitGTR <- optim.pml(fitGTR, model="GTR", optInv=TRUE, optGamma=TRUE,
                    rearrangement = "stochastic", control = pml.control(trace = 0))


#saveRDS(fitGTR, "fitGTR")

```



```{r}
# Handoff to phyloseq ----
# Load all data
#fitGTR <- readRDS("fitGTR")
#otu_table <- read.csv("ES_OTU_table.csv") %>%
#  column_to_rownames(., var = "Sample")
envdata <-  read.csv("sample_data.csv", row.names = 1, header = TRUE)
envdata$Depth2 <- as.factor(envdata$Depth)

# Change the row names of the ASV table to match the environmental table 
map_df <- as.data.frame(read.csv("sample_name_map.csv"))
rownames(seqtab_nochim) <- map_df$new[match(rownames(seqtab_nochim), map_df$old)]

#Final phyloseq creation
physeq.es0 <- phyloseq(otu_table(seqtab_nochim, taxa_are_rows=FALSE), 
                    tax_table(taxa.silva),
                    sample_data(envdata),
                    phy_tree(fitGTR$tree))

#remove low abundance taxa
taxa_sums_total <- taxa_sums(physeq.es0)

physeq.es <- prune_taxa(taxa_sums_total > 10, physeq.es0)

ntaxa(physeq.es0)
ntaxa(physeq.es)
#removes ~500 ASVs with low abundance

new.names <- paste0("ASV_", seq(ntaxa(physeq.es))) # define new names ASV_1, ASV_2, ...
seqs <- taxa_names(physeq.es) # store sequences
names(seqs) <- new.names # make map from ASV to full sequence
taxa_names(physeq.es) <- new.names # rename 

#saveRDS(physeq.es, "physeq.es")

#Save FASTA
seqs2 <- DNAStringSet(seqs)
#writeXStringSet(seqs2, filepath = "es_illumina_not-trimmed.fasta")

```

```{r}
#modify sample data file
physeq.es<- readRDS("physeq.es")

sample_data(physeq.es) <- sample_data(envdata)
saveRDS(physeq.es, "physeq.es")
```