---
title: "Biological diversity"
author: "Stephen Noell"
date: "01/07/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      message = FALSE, 
                      warning = FALSE,
                      fig.align = "center",
                      fig.width = 10,
                      fig.height = 6)
```


```{r Libraries and seed}

# Libraries
library("tidyverse")       # data wrangling and visualisation
library("phyloseq")        # analysis of microbial communities
library("vegan")           # ecological multivariate analyses
library("microbiome") #for analyzing phyloseq objects
library("viridis") #color package
library("svglite") #for saving SVGs
library("ggpubr") #for saving plots
library("here")            # set the path to the folder
library("ggplot2")
library("ggVennDiagram")
library(forcats)
library(pairwiseAdonis)
library(stringr)

set.seed(57)

```

```{r}
#Load version of physeq object, with contaminants removed but not normalized but also rarefied, for alpha diversity analysis
esf.rar <- readRDS("physeq.es.cont.rar")

#Load version of physeq object, with contaminants removed but not normalized
esf <- readRDS("physeq.es.cont")

#Load version of physeq object, with contaminants removed and relative abundance normalized
esfrel <- readRDS("physeq.es.rel")

```


```{r alpha diversity}
#inspect alpha diversity of samples

#comparing ice thickness (i.e. seasonality)

shannon_plot <- plot_richness(esf.rar, x="Ice.thickness",  measures=c("Shannon"))  +
  geom_boxplot(aes(color = Ice.thickness)) +
  geom_jitter(size = 2) +
  ylab("Shannon\nIndex") +
  scale_color_manual(values=c("#addd8e", "#41b6c4", "#225ea8", "#081d58"),
                      name = "Ice thickness") +
  theme_bw() +
  theme(
    axis.title.y = element_text(size=11),
    axis.text.y  = element_text(size=10),
    axis.ticks.x = element_blank(),
    axis.text.x = element_blank(),
    axis.title.x = element_blank(),
    strip.text = element_blank(),
    legend.position="top") +
  stat_pwc(method = "wilcox_test", p.adjust.method = "BH") +
  guides(color=guide_legend(nrow=2,byrow=TRUE))

shannon_plot

### no sig diffs


```



```{R beta diversity}
#Prepare data for PCoA plotting
unifrac.esf.rel <- phyloseq::distance(esfrel, method = "unifrac")
esf.rel.mds <- ordinate(esfrel, "MDS", distance = unifrac.esf.rel)

# final plot
pcoa.esf.rel<- plot_ordination(esfrel, esf.rel.mds, 
                shape="Ice.thickness") +
  geom_point(aes(fill=Dist.from.bottom), size=5,
             alpha = 0.9) +
  #geom_text(label = rownames(esf.rel.mds$vectors),
   #         nudge_x = 0.01, nudge_y = 0.01, 
    #        check_overlap = T) +
  theme_bw() +
  theme(
    axis.title.x = element_text (size=11),
    axis.text.x  = element_text(size=10),
    axis.title.y = element_text(size=11),
    axis.text.y  = element_text(size=10),
    legend.text = element_text(size=10),
    legend.title = element_text(size=10),
    legend.position = "right",
    legend.box = "vertical") +
  scale_fill_gradientn(colors = c("black", "#ca0020", "white"),
  name = "Dist. from\nbottom (cm)") +
  #scale_fill_viridis_c(option = "rocket", name = "Dist. from bottom (cm)") +
  scale_shape_manual(values = c(21, 22, 23, 24), guide = "none") +
  guides(color = guide_colorbar(title.position = "top"))  # <-- continuous guide

pcoa.esf.rel

###DO
pcoa.esf.rel.d<- plot_ordination(esfrel, esf.rel.mds, 
                shape="Ice.thickness") +
  geom_point(aes(fill=DO), size=5,
             alpha = 0.9) +
  #geom_text(label = rownames(esf.rel.mds$vectors),
   #         nudge_x = 0.01, nudge_y = 0.01, 
    #        check_overlap = T) +
  theme_bw() +
  theme(
    axis.title.x = element_text (size=11),
    axis.text.x  = element_text(size=10),
    axis.title.y = element_text(size=11),
    axis.text.y  = element_text(size=10),
    legend.text = element_text(size=10),
    legend.title = element_text(size=10),
    legend.position = "right",
    legend.box = "vertical") +
  scale_fill_viridis_c(option = "plasma", name = "Dissolved\nOxygen (mg/L)") +
  scale_shape_manual(values = c(21, 22, 23, 24), guide = "none") +
  guides(color = guide_colorbar(title.position = "top"))  # <-- continuous guide

pcoa.esf.rel.d

###Conductivity
pcoa.esf.rel.c<- plot_ordination(esfrel, esf.rel.mds, 
                shape="Ice.thickness") +
  geom_point(aes(fill=Conductivity.mS), size=5,
             alpha = 0.9) +
  #geom_text(label = rownames(esf.rel.mds$vectors),
   #         nudge_x = 0.01, nudge_y = 0.01, 
    #        check_overlap = T) +
  theme_bw() +
  theme(
    axis.title.x = element_text (size=11),
    axis.text.x  = element_text(size=10),
    axis.title.y = element_text(size=11),
    axis.text.y  = element_text(size=10),
    legend.text = element_text(size=10),
    legend.title = element_text(size=10),
    legend.position = "right",
    legend.box = "vertical") +
  scale_fill_viridis_c(option = "mako", name = "Conductivity (mS)") +
  scale_shape_manual(name = "Ice thickness", values = c(21, 22, 23, 24)) +
  guides(color = guide_colorbar(title.position = "top"))  # <-- continuous guide

pcoa.esf.rel.c


```

```{r PERMANOVA}

# Now we will perform a PERMANOVA to see if these groups have different centroids
# This will be complemented with a Permutation test for homogeneity of multivariate dispersions 
# -if this doesn't come back significant we can trust that the PERMANOVA results show
#differences in data centroids and not in data dispersion

# Create a data frame with the environmental data 
envdata <-  read.csv("sample_data.csv", header = TRUE)
unifrac.esf.rel.mat <- as.matrix(unifrac.esf.rel)

# PERMANOVA on ice thickness (i.e., season time point)
adonis2(unifrac.esf.rel ~ Ice.thickness, data = envdata)

#dispersion test
disp.site<- betadisper(unifrac.esf.rel, envdata$Ice.thickness)
permutest(disp.site)

# PERMANOVA on depth
adonis2(unifrac.esf.rel ~ Dist.from.bottom, data = envdata)
disp.site<- betadisper(unifrac.esf.rel, envdata$Dist.from.bottom)
permutest(disp.site)

# We see that Depth, but not Ice thickness, returned a significant PERMANOVA and non-significant permutation test

#stronger correlation when doing mixolimnion vs. chemocline
adonis2(unifrac.esf.rel ~ Depth.bin, data = envdata)
disp.site<- betadisper(unifrac.esf.rel, envdata$Depth.bin)
permutest(disp.site)

```



```{r}
# Make community composition plot to see what's different between the time points
#new colors
pal_barn <- RColorBrewer::brewer.pal(5, "Paired")
pal_barnf <- c("black", pal_barn, "grey")

#plot aggregated

merged_group <- phyloseq::merge_samples(esf,group = "Ice.thickness")
merged_groupa<- transform_sample_counts(merged_group, function(x) x / sum(x) )

# Data for plotting
es_pm <- merged_groupa %>% psmelt() %>%
  select(Sample, Phylum, Abundance) %>%
  group_by(Sample, Phylum) %>%
  summarise(TotalAbundance = sum(Abundance), .groups = "drop") %>%
  dplyr::rename(Abundance = TotalAbundance) %>%
  ungroup(.) %>%
  as.data.frame(.)

es_pm["Phylum"][which(es_pm$Abundance < 0.01), ] <- "Phyla < 1 %"
es_pm$Phylum[is.na(es_pm$Phylum)] = "Unassigned"


#Reorder phyla so <1% is first, then alphabetical
list <- es_pm$Phylum %>%
  unique(.) %>%
  as.data.frame(.) %>%
  filter(., . != "Unassigned") %>%
  filter(., . != "Phyla < 1 %") %>%
  unlist(., use.names=FALSE) %>%
  sort(.)

x <- c("Phyla < 1 %", list, "Unassigned")

es_pm$Phylum <- factor(as.character(es_pm$Phylum), levels=x)
es_pm <- es_pm[order(es_pm$Phylum),]

#plot

plot_phylum_a <- ggplot(es_pm, aes(x = Sample, y = Abundance, fill = Phylum)) +
  geom_bar(stat = "identity") +
  scale_fill_manual(values = pal_barnf) +
  labs(x = "Ice thickness (cm)", y = "Relative\nAbundance") +
  theme(
    strip.text.x = element_text(size = 11),
    legend.title = element_blank(),
    legend.text = element_text(size=10),
    axis.text.x  = element_text(size=11, angle = 90),
    axis.title.x = element_text(size=11),
    axis.text.y  = element_text(size=11),
    axis.title.y = element_text(size=11),
    legend.position = "bottom",
    legend.key.width = unit(0.3, "cm"),
    legend.key.height=unit(0.3,"cm")) +
  coord_cartesian(expand = FALSE) +
  guides(fill = guide_legend(nrow = 4, byrow = TRUE))

plot_phylum_a

es_pm %>% filter(Phylum == "Nanoarchaeota") %>% dplyr::group_by(Sample) %>% summarise(sum(Abundance))

```


```{r community composition - 16S}
# Make community composition plot to see what's different between the time points but also across depths
pal_barn <- RColorBrewer::brewer.pal(8, "Paired")
pal_barnf_p <- c("black", pal_barn)

## Phylum abundance plot
# Data for plotting
es_p <- esfrel %>% psmelt() %>%
  select(Sample, Dist.from.bottom, Ice.thickness, Phylum, Abundance)

# Step 1: find max abundance per Phylum across samples
phylum_max <- es_p %>%
  group_by(Phylum) %>%
  summarise(max_abund = max(Abundance, na.rm = TRUE))

# Step 2: identify "rare" phyla
rare_phyla <- phylum_max$Phylum[phylum_max$max_abund < 0.01]

# Step 3: relabel
es_p <- es_p %>%
  mutate(Phylum = ifelse(Phylum %in% rare_phyla, "Phyla < 1 %", Phylum),
         Phylum = ifelse(is.na(Phylum), "Unassigned", Phylum))

#Reorder phyla so <1% is first, then alphabetical
list <- es_p$Phylum %>%
  unique(.) %>%
  as.data.frame(.) %>%
  filter(., . != "Unassigned") %>%
  filter(., . != "Phyla < 1 %") %>%
  unlist(., use.names=FALSE) %>%
  sort(.)

x <- c("Phyla < 1 %", list, "Unassigned")

es_p$Phylum <- factor(as.character(es_p$Phylum), levels=x)
es_p <- es_p[order(es_p$Phylum),]
es_p$Dist.from.bottom <- as.factor(es_p$Dist.from.bottom)

# Plot 
plot_phylum<- ggplot(es_p, aes(x = Dist.from.bottom, y = Abundance, fill = Phylum)) +
  geom_bar(stat = "identity") +
  scale_fill_manual(values = pal_barnf_p) +
  labs(x = "Distance from bottom  (cm)", y = "Relative\nAbundance") +
  theme(
    strip.text.x = element_text(size = 11),
    legend.title = element_blank(),
    legend.text = element_text(size=10),
    axis.text.x  = element_text(size=11, angle = 90),
    axis.title.x = element_text(size=11),
    axis.text.y  = element_text(size=11),
    axis.title.y = element_text(size=11),
    legend.position = "bottom",
    legend.key.width = unit(0.3, "cm"),
    legend.key.height=unit(0.3,"cm")) +
  coord_cartesian(expand = FALSE) +
  coord_flip() +
  facet_wrap(. ~ Ice.thickness, nrow = 4, scales = "free_y") +
  guides(fill = guide_legend(nrow = 4, byrow = TRUE))

plot_phylum

```

```{r community composition - 16S}
# Make community composition plot to see what's different between the time points but also across depths
## Class abundance plot
# Data for plotting

pal_barn2 <- RColorBrewer::brewer.pal(6, "PuOr")
pal_barn3 <- RColorBrewer::brewer.pal(4, "RdGy")
#pal_barn4 <- RColorBrewer::brewer.pal(3, "Purples")
pal_barnf_c <- c("black", pal_barn2, pal_barn3)

es_c <- esfrel %>% psmelt() %>%
  select(Dist.from.bottom, Ice.thickness, Class, Abundance)

# Step 1: find max abundance per Phylum across samples
class_max <- es_c %>%
  group_by(Class) %>%
  summarise(max_abund = max(Abundance, na.rm = TRUE))

# Step 2: identify "rare" classes
rare_classes <- class_max$Class[class_max$max_abund < 0.01]

# Step 3: relabel
es_c <- es_c %>%
  mutate(Class = ifelse(Class %in% rare_classes, "Class < 1 %", Class),
         Class = ifelse(is.na(Class), "Unassigned", Class))


#Reorder class so <X% is first, then alphabetical
list <- es_c$Class %>%
  unique(.) %>%
  as.data.frame(.) %>%
  filter(., . != "Unassigned") %>%
  filter(., . != "Class < 1 %") %>%
  unlist(., use.names=FALSE) %>%
  sort(.)

x <- c("Class < 1 %", list, "Unassigned")

es_c$Class <- factor(as.character(es_c$Class), levels=x)
es_c <- es_c[order(es_c$Class),]
es_c$Dist.from.bottom <- as.factor(es_c$Dist.from.bottom)

# Plot 
plot_class<- ggplot(es_c, aes(x = Dist.from.bottom, y = Abundance, fill = Class)) +
  geom_bar(stat = "identity") +
  scale_fill_manual(values = pal_barnf_c) +
  labs(x = "Distance from bottom (cm)", y = "Relative\nAbundance") +
  theme(
    strip.text.x = element_text(size = 11),
    legend.title = element_blank(),
    legend.text = element_text(size=10),
    axis.text.x  = element_text(size=11, angle = 90),
    axis.title.x = element_text(size=11),
    axis.text.y  = element_text(size=11),
    axis.title.y = element_text(size=11),
    legend.position = "bottom",
    legend.key.width = unit(0.3, "cm"),
    legend.key.height=unit(0.3,"cm")) +
  coord_cartesian(expand = FALSE) +
  coord_flip() +
  facet_wrap(. ~ Ice.thickness, nrow = 4, scales = "free_y") +
  guides(fill = guide_legend(nrow = 4, byrow = TRUE))

plot_class

```



```{r}
#final plot: no alpha diversity

pcoa1 <- ggarrange(pcoa.esf.rel, pcoa.esf.rel.c,
          labels = c("A", "B"),
          ncol = 2, nrow = 1)

pcoa1


beta_bara <- ggarrange(pcoa.esf.rel.d, plot_phylum_a,
                       labels = c("C", "D"),
                       ncol = 2)

beta_bara

bar2 <- ggarrange(plot_phylum,
                           plot_class,
                          labels = c("E", "F"),
                          ncol = 2, nrow = 1)


bio_div_plot2 <- ggarrange(pcoa1, beta_bara, bar2,
                          labels = c("", "", ""),
                          ncol = 1, nrow = 3,
                          heights = c(0.6, 0.6, 1))

bio_div_plot2

#ggsave("Figure3.png", plot = bio_div_plot2, dpi = 300, width = 240, height = 330, units = "mm")
#ggsave("Figure3.svg", plot = bio_div_plot2, dpi = 300, width = 240, height = 330, units = "mm")

##################
#alpha diversity

#ggsave("FigureS3.png", plot = shannon_plot, dpi = 300, width = 100, height = 100, units = "mm")
#ggsave("FigureS3.svg", plot = shannon_plot, dpi = 300, width = 100, height = 100, units = "mm")


```



```{r}
#look at abundances of photosynthesizers
chlor <- readRDS("physeq.es.chlor")

es.chl.heat <- chlor %>% 
  psmelt()

# aggregate by Order instead of OTU
es.chl.heat.filt <- es.chl.heat %>%
  mutate(Abundance = Abundance * 100) %>%   # percent scale
  group_by(Sample, Dist.from.bottom, Ice.thickness, Order) %>%
  summarise(Abundance = sum(Abundance), .groups = "drop")

es.chl.heat.filt$Dist.from.bottom <- as.factor(es.chl.heat.filt$Dist.from.bottom)

# Plot heatmap
heatmap_order <- ggplot(es.chl.heat.filt, aes(Dist.from.bottom, Order)) +
  geom_tile(aes(fill = Abundance)) +
  theme_bw() +
  scale_fill_distiller('Relative\nabundance (%)', palette = "YlOrRd", direction = 1) +
  ylab("Order") +
  xlab("Distance from bottom (cm)") +
  theme(
    axis.title.x = element_text(size=10),
    axis.text.x  = element_text(size=8),
    axis.title.y = element_blank(),
    axis.text.y  = element_text(size=8),
    strip.text.x = element_text(size=8),
    legend.text = element_text(size=8),
    legend.title = element_text(size=10),
    legend.position = "bottom") +
  scale_x_discrete(guide = guide_axis(angle = 90)) +
  facet_grid(.~factor(Ice.thickness), scales = "free", space = "free")

heatmap_order

#ggsave("FigureS1.svg", plot = heatmap_order, dpi = 300, width = 200, height = 100, units = "mm")
#ggsave("FigureS1.png", plot = heatmap_order, dpi = 300, width = 200, height = 100, units = "mm")

chlor2 <- es.chl.heat.filt %>% dplyr::group_by(Order) %>% dplyr::filter(Order == "Chloroplast") %>% 
  ungroup(.) %>% as.data.frame(.)

max(chlor2$Abundance)

```


```{r}
chlor <- readRDS("physeq.es.chlor")

#Look at the abundance profile of whatever ASVs you want to using heatmaps
es.chl.heat <- chlor %>%
  psmelt()

es.chl.heat.filt <- es.chl.heat

es.chl.heat.filt <- es.chl.heat.filt %>%
  mutate(Abundance2 = (Abundance*100)) %>%
  select(-Abundance) %>%
  dplyr::rename(Abundance = Abundance2)

es.chl.heat.filt$Dist.from.bottom <- as.factor(es.chl.heat.filt$Dist.from.bottom)

es.chl.heat.filt$ASV_Phylum = paste(
  es.chl.heat.filt$Phylum, 
                                       es.chl.heat.filt$Class,
                                   es.chl.heat.filt$Order, 
                                   #es.chl.heat.filt$Family,
                                   #es.chl.heat.filt$Genus, 
                                   es.chl.heat.filt$OTU)

heatmap_asv <- ggplot(es.chl.heat.filt, aes(Dist.from.bottom, ASV_Phylum)) +
  geom_tile(aes(fill = Abundance)) +
  theme_bw() +
  scale_fill_distiller('Relative\nabundance (%)', palette = "YlOrRd", direction = 1) +
  ylab("OTU") +
  xlab("Distance from bottom (cm)") +
  theme(
    axis.title.x = element_text (size=10),
    axis.text.x  = element_text(size=8),
    axis.title.y = element_blank(),
    axis.text.y  = element_text(size=8),
    strip.text.x = element_text(size = 8),
    legend.text = element_text(size=8),
    legend.title = element_text(size=10),
    legend.position = "bottom") +
    scale_x_discrete(guide = guide_axis(angle = 90)) +
  facet_grid(.~factor(Ice.thickness), scales = "free",
             space='free')

heatmap_asv


```

```{r}
#nanoarchaea heatmap
########
library(pheatmap)
library(RColorBrewer)          # for color palettes

nanos <- readRDS("physeq.es.rel")

#Look at the abundance profile of whatever ASVs you want to using heatmaps
es.nan.heat <- nanos %>%
  psmelt()

es.nan.heat$Phylum[is.na(es.nan.heat$Phylum)] = "Unassigned"
es.nan.heat$Genus[is.na(es.nan.heat$Genus)] = ""

es.nan.heat.filt <- es.nan.heat %>%
  dplyr::filter(., Phylum == "Nanoarchaeota")

#es.nan.heat.filt <- es.nan.heat.filt %>%
 # mutate(Abundance2 = (Abundance*100)) %>%
  #select(-Abundance) %>%
  #dplyr::rename(Abundance = Abundance2)

es.nan.heat.filt$Dist.from.bottom <- as.factor(es.nan.heat.filt$Dist.from.bottom)

es.nan.heat.filt$ASV_Phylum = paste(
  es.nan.heat.filt$Phylum, 
                                       es.nan.heat.filt$Class,
                                   es.nan.heat.filt$Order, 
                                   #es.nan.heat.filt$Family,
                                   #es.nan.heat.filt$Genus, 
                                   es.nan.heat.filt$OTU)

es.nan.heat.filt$News = paste(
  es.nan.heat.filt$Ice.thickness, es.nan.heat.filt$Dist.from.bottom)

phyla_abund_m <- es.nan.heat.filt %>%
  dplyr::select(ASV_Phylum, Abundance, News) %>%
  dplyr::filter(Abundance > 0.1)

p_a_f <- reshape2::dcast(phyla_abund_m, ASV_Phylum ~ News, value.var = "Abundance", fill = 0) %>%
  data.frame(., row.names = 1)

p_a_f_m <- as.matrix(p_a_f)

p_a_f_m[is.na(p_a_f_m)] <- 0
colnames(p_a_f_m) <- gsub("^X", "", colnames(p_a_f_m))   # remove leading X
colnames(p_a_f_m) <- gsub("\\.", " ", colnames(p_a_f_m)) # replace dots with spaces

###add annotation heatmap
envdata <-  read.csv("sample_data.csv", header = TRUE)
envdata$News = paste(
  envdata$Ice.thickness, envdata$Dist.from.bottom)

# Make sure sample names match the column names in p_a_f_m
env_ann <- envdata %>%
  dplyr::select(News, DO, Conductivity.mS, Temperature)

# set rownames = sample names to match heatmap columns
rownames(env_ann) <- env_ann$News
env_ann$News <- NULL

# annotation colors (continuous variables will get gradients)
ann_colors <- list(
  DO = colorRampPalette(brewer.pal(9, "Blues"))(100),
  Conductivity.mS = colorRampPalette(brewer.pal(9, "Greens"))(100),
  Temperature = colorRampPalette(brewer.pal(9, "Oranges"))(100)
)

pheatmap(p_a_f_m,
         color = colorRampPalette(brewer.pal(8, "RdPu"))(25),
         annotation_col = env_ann,
         annotation_colors = ann_colors,
         fontsize = 7,
         width = 7,
         height = 4,
         filename = "FigureS4.png")

#Get relative abundance at each ice thickness of Nanoarchaeota

nano_num <- es.nan.heat.filt %>%
  dplyr::group_by(Ice.thickness) %>%
  summarise(sum(Abundance))

# Divide each by num of samples: 13cm: 5; 28 and rest: 6
nano_num$divide <- c(5, 6, 6, 6)

nano_num2 <- nano_num %>%
  dplyr::mutate(prop = `sum(Abundance)` / divide)

nano_num2$prop

```


```{r}
#generate Table S3
library(Biostrings)
otu_table <- data.frame(otu_table(esfrel)) %>%
  rownames_to_column(var = "Sample")

library(data.table)
otu_long <- data.table::transpose(setDT(otu_table), make.names = 'Sample', 
                                  keep.names = 'Sample') %>%
  as.data.frame(.) %>%
  dplyr::rename(ASV = 1)

tax_table <- data.frame(tax_table(esfrel)) %>%
  rownames_to_column(var = "ASV")

# load the sequences from the file
fasta <- readDNAStringSet("es_illumina_not-trimmed.fasta")

# Extract ASV names and sequences
otu_seqs <- tibble(
  ASV = names(fasta),
  Sequence = as.character(fasta)
)

# Join long ASV table to taxonomy and sequences
full_df <- tax_table %>%
  left_join(otu_long, by = "ASV") %>%
  left_join(otu_seqs, by = "ASV")

#write.csv(full_df, "Table S3.csv")

# Create DNAStringSet object
sequences <- DNAStringSet(full_df$Sequence)
names(sequences) <- full_df$ASV

# Write FASTA file
#writeXStringSet(sequences, "es_illumina_trimmed_ASVs_nochim.fasta")

#fixing genbank submission

genb <- read.table("source_modifiers.tsv", sep = "\t", header = TRUE, stringsAsFactors = FALSE)

full_df <- full_df %>%
  dplyr::rename("Sequence_ID" = "ASV")

genb2 <- left_join(genb, full_df, by = "Sequence_ID")

genb3 <- genb2 %>%
  dplyr::mutate(Organism = case_when(
    Kingdom == "Bacteria" ~ "uncultured aquatic bacterium",
    Kingdom == "Archaea" ~ "uncultured aquatic archaeon",
    TRUE ~ NA_character_  # for any other values
  )) %>%
  dplyr::select(c(1:6))

#write.table(genb3, "source_modifiers2.tsv", sep = "\t", row.names = FALSE, quote = FALSE)
```
